lib/src/component/error_boundary.dart

```diff
@@ -1,14 +1,72 @@
+///
+/// > __Why this is here__
+/// >
+/// > In order to release react-dart 5.0.0 _(which upgrades to ReactJS 16)_
+///   without depending on Dart 2 / `Component2` (coming in react-dart 5.1.0) / `UiComponent2` (coming in over_react 3.1.0) -
+///   and all the new lifecycle methods that those expose, we need to ensure that - at a minimum - the `componentDidCatch`
+///   lifecycle method is handled by components wrapped in our [ErrorBoundary] component so that the behavior of
+///   an application when a component within a tree throws - is the same as it was when using ReactJS 15.
+/// >
+/// > Otherwise, the update to react-dart 5.0.0 / over_react 3.0.0 will result in consumer apps rendering completely
+///   "blank" screens when their trees unmount as a result of a child component throwing an error.
+///   This would be unexpected, unwanted - and since we will not add a Dart js-interop layer around `componentDidCatch`
+///   until react-dart 5.1.0 / over_react 3.1.0 - unmanageable for consumers.
+///
+/// __This will be removed in over_react 3.1.0__ once [ErrorBoundaryComponent] is extending from `UiStatefulComponent2`
+/// which will ensure that the [ErrorBoundaryComponent.componentDidCatch] lifecycle method has real js-interop bindings
+/// via react-dart 5.1.0's `Component2` base class.
+///
+/// TODO: Remove in 3.1.0
+final ReactElement Function([Map props, List children]) _jsErrorBoundaryComponentFactory = (() {
+  var componentClass = React.createClass(jsifyAndAllowInterop({
+    'displayName': 'JsErrorBoundary',
+    'render': allowInteropCaptureThis((jsThis) {
+      final jsProps = js_util.getProperty(jsThis, 'props');
+      return js_util.getProperty(jsProps, 'children');
+    }),
+    'componentDidCatch': allowInteropCaptureThis((jsThis, error, info) {
+      final jsProps = js_util.getProperty(jsThis, 'props');
+      js_util.getProperty(jsProps, 'onComponentDidCatch')(error, info);
+    }),
+  }));
+
+  // Despite what the ReactJS docs say about only needing _either_ componentDidCatch or getDerivedStateFromError
+  // in order to define an "error boundary" component, that is not actually the case.
+  //
+  // The tree will never get re-rendered after an error is caught unless both are defined.
+  // ignore: argument_type_not_assignable
+  js_util.setProperty(componentClass, 'getDerivedStateFromError', allowInterop((_) => js_util.newObject()));
+
+  var reactFactory = React.createFactory(componentClass);
+
+  return ([Map props = const {}, List children = const []]) {
+    return reactFactory(jsifyAndAllowInterop(props), listifyChildren(children));
+  };
+})();
+
 // TODO: Need to type the second argument once react-dart implements bindings for the ReactJS "componentStack".
-typedef _ComponentDidCatchCallback(Error error, /*ComponentStack*/dynamic componentStack);
+// TODO: Will it be possible to make the first argument typed as a real `Error` instance again in Dart 2?
```

Comment:
No, since `Exception` does not extend from `Error`, and also the error thrown could be neither of those types.

`error` should stay `dynamic`, which is what `Future.catchError` and `runZoned` use.